pkg/querier/queryrange/querysharding.go

```diff
@@ -0,0 +1,246 @@
+package queryrange
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/loki/pkg/loghttp"
+	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/logql/marshal"
+)
+
+var nanosecondsInMillisecond = int64(time.Millisecond / time.Nanosecond)
+
+// NewQueryShardMiddleware creates a middleware which downstreams queries after AST mapping and query encoding.
+func NewQueryShardMiddleware(
+	logger log.Logger,
+	confs queryrange.ShardingConfigs,
+	minShardingLookback time.Duration,
+	middlewareMetrics *queryrange.InstrumentMiddlewareMetrics,
+	shardingMetrics *logql.ShardingMetrics,
+) queryrange.Middleware {
+
+	noshards := !hasShards(confs)
+
+	if noshards {
+		level.Warn(logger).Log(
+			"middleware", "QueryShard",
+			"msg", "no configuration with shard found",
+			"confs", fmt.Sprintf("%+v", confs),
+		)
+		return queryrange.PassthroughMiddleware
+	}
+
+	mapperware := queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return newASTMapperware(confs, next, logger, shardingMetrics)
+	})
+
+	return queryrange.MiddlewareFunc(func(next queryrange.Handler) queryrange.Handler {
+		return &shardSplitter{
+			MinShardingLookback: minShardingLookback,
+			shardingware: queryrange.MergeMiddlewares(
+				queryrange.InstrumentMiddleware("shardingware", middlewareMetrics),
+				mapperware,
+			).Wrap(next),
+			now:  time.Now,
+			next: queryrange.InstrumentMiddleware("sharding-bypass", middlewareMetrics).Wrap(next),
+		}
+	})
+
+}
+
+func newASTMapperware(
+	confs queryrange.ShardingConfigs,
+	next queryrange.Handler,
+	logger log.Logger,
+	metrics *logql.ShardingMetrics,
+) *astMapperware {
+
+	return &astMapperware{
+		confs:   confs,
+		logger:  log.With(logger, "middleware", "QueryShard.astMapperware"),
+		next:    next,
+		metrics: metrics,
+	}
+}
+
+type astMapperware struct {
+	confs   queryrange.ShardingConfigs
+	logger  log.Logger
+	next    queryrange.Handler
+	metrics *logql.ShardingMetrics
+}
+
+func (ast *astMapperware) Do(ctx context.Context, r queryrange.Request) (queryrange.Response, error) {
+	conf, err := ast.confs.GetConf(r)
+	// cannot shard with this timerange
+	if err != nil {
+		level.Warn(ast.logger).Log("err", err.Error(), "msg", "skipped AST mapper for request")
+		return ast.next.Do(ctx, r)
+	}
+
+	shardedLog, ctx := spanlogger.New(ctx, "shardedEngine")
+	defer shardedLog.Finish()
+
+	ng, err := logql.NewShardedEngine(logql.EngineOpts{}, int(conf.RowShards), DownstreamHandler{ast.next}, ast.metrics, shardedLog)
```

Comment:
Should `logql.EngineOpts{}` be configurable ? since it has a timeout ?

Is it not possible to have this created only once ? It would make the metrics registration more simple also through Prometheus register only.

I don't know if we want to log all of this into a span ? Unless it has no performance impact, I would be careful may be @joe-elliott or @annanay25 can answer this question ? I'm genuinely not sure.

Reply:
I'm not sure I understand the concern. I see that @owen-d is opening a span and closing it, but I don't see an excessive number of tags/logs added to the span.

Reply:
Since logs are not indexed they have a much lower overhead compared to tags, so as long as we are creating a new instance of the `SpanLogger` for every request (I couldn't confirm that from first look), I think it should work just fine

Reply:
@joe-elliott the logger is used then as a dependency in NewShardedEngine. But if this is not indexed that should be fine, not sure how much it will log though I guess we will tweak that once we have it running.

Reply:
Ah, I see now.  Thanks @annanay25/@cyril.  

Yeah, passing that span logger in as a log.Logger is kind of dangerous.  My primary concern is that the side effects of using that log are extremely non-obvious.  A developer may add a bunch of debug logs without realizing they're offloading stacks of data to the tracing backend.

Reply:
Great points. I'll stick to deriving this via context